tests/test_wires.py

```diff
@@ -110,6 +110,21 @@ def test_equality(self):
         assert wires1 != wires2
         assert wires1 == wires3
 
+    def test_less_an_greater_than(self):
+        """Tests that we can compare Wires objects with the '>' and '<' operators."""
+
+        wires1 = Wires([1, 'q2', 3])
+        wires2 = Wires(['q1', 2, 1])
+        wires3 = Wires(['a', 1, 'b'])
+        assert wires1 < wires2
+        assert wires3 < wires2
+        assert wires1 < wires2
+        assert wires2 > wires1
+        assert wires2 > wires3
+        assert wires2 > wires1
+        assert not wires1 < ['q1', 2, 1]
+        assert not wires2 > [1, 'q2', 3]
```

Comment:
Unfortunately, I had no idea what to expect from `>` and `<`. :thinking:.

Reply:
A bigger issue is that Python orders sequences (like strings and lists of ints) in a way that I think is incompatible with what you want to do here.

Consider the following example (copied from my Python terminal):
```
>>> [0,1]<[1,2]
>>> True  # seems legit
>>> [0,1]<[1,-1]
>>> True  # ok, maybe something wrong with negative numbers?
>>> [0,1]<[1,0]
>>> True  # :thinking: 
>>> [0,1]<[1,"hi"]
>>> True  # no idea if this should be true or false
```
Note that this could lead to situations where `assert wires1 > wires2` and `assert wires2 > wires1` both hold

Reply:
https://stackoverflow.com/questions/30964157/python-inequality-operators-comparing-lists

Reply:
This functionality would only be for the `groupby` function, so the goal is to have *any* comparison that is reproducible. But you are right, there might be side effects that would be non-intuitive. An alternative I thought of was to implement a comparison of string representations? Or to rewrite the entire eigenvalue sorting system in PL, which I'd rather avoid.

Reply:
Sadly, string comparison works the same as list comparisons.

The current implementation might provide self-consistent comparisons (i.e., the value of a comparison is unambiguously `True` or `False` every time), but it doesn't provide any self-consistent *ordering*. Which one of these do we need for eigenvalue sorting? (I'd assume the second)

Reply:
I couldn't quite follow the above, is it possible to have a quick summary of why we want the ability to place an ordering?

Reply:
I think we might have to revisit `Tensor.eigvals` and `Tensor.matrix` in an ADR, my gut feeling is that the current implementation should be rethought.

@trbromley I believe this context is this line, from the `Tensor.eigvals` property:

```python
        # Sorting the observables based on wires, so that the order of
        # the eigenvalues is correct
        obs_sorted = sorted(self.obs, key=lambda x: x.wires)
```

This sorting is to ensure that the returned eigenvalues are in the correct order for the tensor observable.

Essentially: we need the returned eigenvalues to be in the same order as they would be if we manually diagonalized the tensor observable in the computational basis (e.g., the returned eigenvalues must be in the same order as `diag(Lambda)`, where `Lambda = S.H.S^ct`, `H` is the observable, and `S` is some unitary change of basis operation that diagonalizes `H`).

Currently, we attempt to sort the observables in the tensor product by wire location, and then directly access the eigenvalues for each term in the product. This is to avoid doing any expensive diagonalization. But there are a lot of assumptions baked in here; it will work fine for
```python
qml.PauliZ(0) @ qml.Hermitian(A, wires=[2, 4]) @ qml.PauliX(1)
```
but (I'm pretty sure?) isn't well defined for something like

```python
qml.PauliZ(0) @ qml.Hermitian(A, wires=[2, 4]) @ qml.PauliX(3)
```

Reply:
Thinking about this some more: we know that the diagonalization operation is given by

```python
def diagonalizing_gates(self):
    diag_gates = []
    for o in self.obs:
        diag_gates.extend(o.diagonalizing_gates())

    return diag_gates
```

Perhaps we can just modify this method so that we can keep track of the observable order, and make sure to use the _same_ order in `Tensor.eigvals`.

Reply:
I think I let myself get overly confused here. After further discussion with @mariaschuld, I think this method should give us a way to define a total order whenever we need it